pyro/distributions/asymmetriclaplace.py

```diff
@@ -74,3 +77,120 @@ def variance(self):
+    the limit of ``softness → 0``, this converges in distribution to the
+    :class:`AsymmetricLaplace` distribution.
+
+    This is equivalent to the sum of three random variables ``z - u + v`` where::
+
+        z ~ Normal(loc, scale * softness)
+        u ~ Exponential(1 / (scale * asymmetry))
+        v ~ Exponential(scale / asymmetry)
+
+    This is also equivalent the sum of two random variables ``z - a`` where::
+
+        z ~ Normal(loc, scale * softness)
+        a ~ AsymmetricLaplace(0, scale, asymmetry)
+
+    :param loc: Location parameter, i.e. the mode.
+    :param scale: Scale parameter = geometric mean of left and right scales.
+    :param asymmetry: Square of ratio of left to right scales. Defaults to 1.
+    :param softness: Scale parameter of the Gaussian smoother. Defaults to 1.
+    """
+    arg_constraints = {"loc": constraints.real,
+                       "scale": constraints.positive,
+                       "asymmetry": constraints.positive,
+                       "softness": constraints.positive}
+    support = constraints.real
+    has_rsample = True
+
+    def __init__(self, loc, scale, asymmetry=1.0, softness=1.0, *, validate_args=None):
+        self.loc, self.scale, self.asymmetry, self.softness = broadcast_all(
+            loc, scale, asymmetry, softness,
+        )
+        super().__init__(self.loc.shape, validate_args=validate_args)
+
+    @lazy_property
+    def left_scale(self):
+        return self.scale * self.asymmetry
+
+    @lazy_property
+    def right_scale(self):
+        return self.scale / self.asymmetry
+
+    @lazy_property
+    def soft_scale(self):
+        return self.scale * self.softness
+
+    def expand(self, batch_shape, _instance=None):
+        new = self._get_checked_instance(AsymmetricLaplace, _instance)
+        batch_shape = torch.Size(batch_shape)
+        new.loc = self.loc.expand(batch_shape)
+        new.scale = self.scale.expand(batch_shape)
+        new.asymmetry = self.asymmetry.expand(batch_shape)
+        new.softness = self.softness.expand(batch_shape)
+        super(AsymmetricLaplace, new).__init__(batch_shape, validate_args=False)
+        new._validate_args = self._validate_args
+        return new
+
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+
+        # Standardize.
+        x = (value - self.loc) / self.scale
+        L = self.asymmetry
+        R = self.asymmetry.reciprocal()
+        S = self.softness
+        SS = S * S
+        S2 = S * math.sqrt(2)
+        Lx = L * x
+        Rx = R * x
+
+        # This is the sum of two integrals which are proportional to:
+        # left = Integrate[e^(-t/L - ((x+t)/S)^2/2)/sqrt(2 pi)/S, {t,0,Infinity}]
+        #      = 1/2 e^((2 L x + S^2)/(2 L^2)) erfc((L x + S^2)/(sqrt(2) L S))
+        # right = Integrate[e^(-t/R - ((x-t)/S)^2/2)/sqrt(2 pi)/S, {t,0,Infinity}]
+        #       = 1/2 e^((S^2 - 2 R x)/(2 R^2)) erfc((S^2 - R x)/(sqrt(2) R S))
```

Comment:
the approach sounds reasonable to me. I haven't checked the details but we can trust the tests I think.